ERA5 interpolation at the prime meridian

einola opened the pull request:
# ERA5 interpolation at the prime meridian

Fixes #888 

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [x] Documented the feature by providing worked example
- [x] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---

There was a small bug in handling the interpolation of ERA5 fields. In ERA5, there is a row of values at the north pole, but you can't define the wind direction there using a lat/lon coordinate system. To address this, we wanted to copy the row south of the pole to this pole-row, as a crude fix, but the code copied the column along the prime meridian instead, which led to strange results there.

I've fixed this, so the copying happens as intended. It's not a perfect solution, but it's good enough for now.

---
# Test Description

Run the old and new versions of the era5_topaz4_maker.py script, and confirm that in the new version, the field is continuous across the prime meridian, while the old version produced a jump there.

---
# Documentation Impact
None

---
# Other Details

None

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [x] File dates have been updated to reflect modification date
- [x] This change conforms to the conventions described in the README

einola committed:
Fix a bug in latitude handling

We wanted to set the values at the pole to be the same as the row below,
but this wasn't correctly coded. The error is corrected now. Arguably,
we should be doing something more fancy, but let's see if this is good
enough for now.

einola committed:
Add a rotate_pole_to_greenland helper function

I added the ParametricMesh::RotatePoleToGreenland to interpolators.py as
rotate_pole_to_greenland for testing purposes. It may be nice to have
in the future, so I'll let it tag along here.

einola commented:
I branched this off issue880_topaz_interpolation, so this PR should be merged once #885 is merged.

joewallwork commented on run/era5_topaz4_maker.py:
```diff
@@ -211,12 +211,12 @@ def topaz4_source_file_name(unix_time, path):
-                    u_data_source[:, -1] = u_data_source[:, -2]
-                    v_data_source[:, -1] = u_data_source[:, -2]
+                    u_data_source[-1, :] = u_data_source[-2, :]
+                    v_data_source[-1, :] = u_data_source[-2, :]
```
I know it's not strictly related to the lines changed but these floating point equality checks are bug-prone. Would recommend `if np.isclose(abs(source_lats[0]), 90.):` and similarly for the other condition.

joewallwork commented on run/interpolators.py:
```diff
@@ -112,3 +112,32 @@ def rotate_velocities(u, v, angle):
+    y = np.cos(latr) * np.sin(lonr)
+    z = np.sin(latr)
+
+    aw = 40.0 * np.pi / 180.0
```
Could use 
```suggestion
    aw = np.deg2rad(40.0)
```
for consistency.

joewallwork commented on run/interpolators.py:
```diff
@@ -112,3 +112,32 @@ def rotate_velocities(u, v, angle):
+    y1 = np.sin(aw) * x + np.cos(aw) * y
+    z1 = z
+
+    bw = 15.0 * np.pi / 180.0
```
Also
```suggestion
    bw = np.deg2rad(15.0)
```

joewallwork commented on run/interpolators.py:
```diff
@@ -112,3 +112,32 @@ def rotate_velocities(u, v, angle):
+def rotate_pole_to_greenland(lat, lon):
+    """
+    Rotates the mesh such that the singularities are in Greenland / Antarctica at 75N / 40W and 75S / 140E
+    This is a copy of ParametricMesh::RotatePoleToGreenland in ParametricMesh.cpp. It's here for testing purposes only.
```
```suggestion
    This is a copy of ParametricMesh::RotatePoleToGreenland in dynamics/src/include/ParametricMesh.hpp. It's here for testing purposes only.
```

joewallwork reviewed (approved):
I can confirm that the formulae are consistent with those used in `ParametricMesh.hpp`. Here are a few optional suggestions.

einola committed:
Use numpy's deg2rad

Use np.deg2rad instead of np.pi/180 to change from degrees to radians. A bit more transparent and easy to read. Suggestion from Joe on github.

Co-authored-by: Joe Wallwork <22053413+jwallwork23@users.noreply.github.com>

einola committed:
Correct reference in helper function

Correctly refer to ParametricMesh.hpp (not .cpp) with a full path. Suggestion from Joe on github.

Co-authored-by: Joe Wallwork <22053413+jwallwork23@users.noreply.github.com>

einola committed:
Don't compare a float using ==

To check if the first or last row contains the pole we used "== 90",
which could, strictly speaking, return a false even when it should be
true if the value is very close to 90, but not 90, due to machine
imprecision. So, knowing that the ERA5 latitudes are spaced at 0.25
degrees, I changed this to "> 89.8". The fields produced look the same
after the change (as they should.

einola committed:
Prefer np.isclose to >

It's easier to read.
